### PR TITLE
Create endpoint for total adherent count + by gender

### PR DESCRIPTION
**Pull request by illusionOfParadise**

### CODE REVIEW

File: src/Controller/Api/ReferentController.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        $count = $adherentRepository->countByGenderAndTotal();
+
+        return new JsonResponse([
+            'male' => self::calculatePercentage($count['male'], $count['total']),
+            'female' => self::calculatePercentage($count['female'], $count['total']),
+            'total' => $count['total'],
+        ]);
+    }
+
+    private static function calculatePercentage(int $nb, int $total): int
```
Comment (B-Galati) [inline src/Controller/Api/ReferentController.php]: No need for `static` no ?

File: tests/Controller/Api/ReferentControllerTest.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+use AppBundle\DataFixtures\ORM\LoadUserData;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Tests\AppBundle\Controller\ControllerTestTrait;
+use Tests\AppBundle\SqliteWebTestCase;
+
+/**
+ * @group functional
+ */
+class ReferentControllerTest extends SqliteWebTestCase
```
Comment (B-Galati) [inline tests/Controller/Api/ReferentControllerTest.php]: Use Behat instead please.
It's more readable.

File: src/Repository/AdherentRepository.php [hunk 3]
```diff
@@ -512,4 +513,16 @@ public function findAdherentsUuidByEmailAddress(string $emailAddress): array
...
+    public function countByGenderAndTotal(): array
+    {
+        $qb = $this->createQueryBuilder('a')
+            ->select('COUNT(a.id) AS total, SUM(case WHEN a.gender = :female then 1 else 0 end) AS female, SUM(case WHEN a.gender = :male then 1 else 0 end) AS male')
+            ->andWhere('a.adherent = 1')
+            ->setParameter('female', Genders::FEMALE)
+            ->setParameter('male', Genders::MALE)
+        ;
+
+        return $qb->getQuery()->getResult()[0];
```
Comment (B-Galati) [inline src/Repository/AdherentRepository.php]: We can call `getSingleResult()` instead.

File: src/Repository/AdherentRepository.php [hunk 4]
```diff
@@ -512,4 +513,16 @@ public function findAdherentsUuidByEmailAddress(string $emailAddress): array
             return $uuid->toString();
         }, array_column($query->getArrayResult(), 'uuid'));
     }
+
+    public function countByGenderAndTotal(): array
+    {
+        $qb = $this->createQueryBuilder('a')
+            ->select('COUNT(a.id) AS total, SUM(case WHEN a.gender = :female then 1 else 0 end) AS female, SUM(case WHEN a.gender = :male then 1 else 0 end) AS male')
```
Comment (B-Galati) [inline src/Repository/AdherentRepository.php]: I am concerned with the performances of this query. 
Perhaps @atamano can try it with prod DB?

Comment (ottaviano) [reply src/Repository/AdherentRepository.php]: I think it'll be better with GROUP BY
ex:
1) `createQueryBuilder('a', 'gender')` => will index the result array with `gender` values
2) `addSelect('a.gender')` => need to be in selected column for using it in index of result array 
3) `addSelect('COUNT(a.id) AS total')`
4) `WHERE adherent = 1 GROUP BY gender`

File: src/Controller/Api/ReferentController.php (hunk 1 above)
Comment (illusionOfParadise) [reply src/Controller/Api/ReferentController.php]: No.

File: tests/Controller/Api/ReferentControllerTest.php (hunk 2 above)
Comment (illusionOfParadise) [reply tests/Controller/Api/ReferentControllerTest.php]: Done.

File: src/Repository/AdherentRepository.php (hunk 4 above)
Comment (illusionOfParadise) [reply src/Repository/AdherentRepository.php]: Done.

File: src/Repository/AdherentRepository.php (hunk 3 above)
Comment (illusionOfParadise) [reply src/Repository/AdherentRepository.php]: No more need after refactoring https://github.com/EnMarche/en-marche.fr/pull/2358/files#diff-e7a7adb167b1d2e64b02883d2be7eff1R518.

File: src/Repository/AdherentRepository.php [hunk 5]
```diff
@@ -512,4 +512,15 @@ public function findAdherentsUuidByEmailAddress(string $emailAddress): array
             return $uuid->toString();
         }, array_column($query->getArrayResult(), 'uuid'));
     }
+
+    public function countByGender(): array
+    {
+        $qb = $this->createQueryBuilder('a', 'a.gender')
+            ->select('a.gender, COUNT(a.id) AS count')
+            ->andWhere('a.adherent = 1')
```
Comment (ottaviano) [inline src/Repository/AdherentRepository.php]: `where`

File: src/Repository/AdherentRepository.php [hunk 6]
```diff
@@ -512,4 +512,15 @@ public function findAdherentsUuidByEmailAddress(string $emailAddress): array
...
+            ->groupBy('a.gender')
+        ;
+
+        return $qb->getQuery()->getArrayResult();
```
Comment (ottaviano) [inline src/Repository/AdherentRepository.php]: IMO `$qb` is useless, you can chain the calls

File: src/Controller/Api/ReferentController.php [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        $count = $adherentRepository->countByGender();
+
+        array_walk($count, function (&$item) {
+            $item = (int) $item['count'];
+        });
+
+        $total = array_sum($count);
+
+        return new JsonResponse([
+            'male' => $this->calculatePercentage($count[Genders::MALE], $total),
```
Comment (ottaviano) [inline src/Controller/Api/ReferentController.php]: what about :
```
$count = $adherentRepository->countByGender();
array_walk($count, function (&$item) {
    $item = (int) $item['count'];
});

$total = array_sum($count);

array_walk($count, function (&$item) use ($total) {
    $item = $this->calculatePercentage($item, $total);
});

$count['total'] = $total;

return new JsonResponse($count);
```

File: src/Controller/Api/AdherentsController.php [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        $count = $adherentRepository->countByGender();
+
+        array_walk($count, function (&$item) {
+            $item = (int) $item['count'];
+        });
+
+        $total = array_sum($count);
+
+        return new JsonResponse([
+            'male' => $this->calculatePercentage($count[Genders::MALE], $total),
```
Comment (ottaviano) [inline src/Controller/Api/AdherentsController.php]: what about:
```
$count = $adherentRepository->countByGender();
array_walk($count, function (&$item) {
    $item = (int) $item['count'];
});

$total = array_sum($count);

array_walk($count, function (&$item) use ($total) {
    $item = $this->calculatePercentage($item, $total);
});

$count['total'] = $total;

return new JsonResponse($count);
```

File: features/api/adherents.feature [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+  I should be able to acces adherents API data
+
+  Background:
+    Given the following fixtures are loaded:
+      | LoadUserData      |
+      | LoadAdherentData  |
+
+  Scenario: As a non logged-in user I can not access the adherents count information
+    When I am on "/api/adherents/count"
+    Then the response status code should be 200
```
Comment (B-Galati) [inline features/api/adherents.feature]: In another PR, we should change this behavior to answer with 401 on /api route (like we did on data-api)

**Pull request merged**